### ISSUE TITLE
calling generic function declared at base class with union type cause unexpected compile error

### ISSUE THREAD

**Issue by wenxiao** [microsoft/TypeScript#49796]
# Bug Report

Calling generic function declared at base class with union type cause unexpected compile error.

### 🔎 Search Terms

Union generic, keyof this

### 🕗 Version & Regression Information

TypeScript version: 4.7.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/dev/bug-workbench/?ts=4.7.4#code/KYDwDg9gTgLgBAbwL4G4CwAoTBjANgQwGdC4AhI4AYQOMUzgbgDMBXAO2wB4BpOUGYGwAmJANbAAnhCZwYACwCWhAHwAKUQC443AJSIkmA1gx4iJAIJ8QA4SXKEqNEgnqN8cALxwAjOgyM4ACNPOAAmPyMcJzIrGxEyCmozOn9GYK8AZj8A7BCAFgjMTFYOGAUINmZ2LgAVABptWMF48SkZGrUANy16uE1tPWQijFxgeHcvNmAAdzhzVR0-UfHArUsAHxjJmbmFuCJ9tgk-YurVfAaAcnxLxdOOc6vA278Aeleqh4u4S+wXuHecGwEAAtmAFKM+FAoNBhoCSthzoErjdFgCPsCwRDgFCYVB7oj8Mifs87sZ8AA6BGqa4vTCU6mXUlvD4Ms6-f6AzHgyHAaGw4yAolU9molDooGgnk4vl4+mBEUPJmcjFS7G46BwAByAHkanAAKIADQACgbKDUDQARTBAA)

### 💻 Code

```tsexport {};

class BaseClass {
    func<K extends keyof this>(k: K) {}
}

class A extends BaseClass {
    a = 1;
    b = 2;
}

class B extends BaseClass {
    b = 3;
    c = 4;
}

function func<T, K extends keyof T>(v: T, k: K) {}

let a = new A();
let ab: A | B = new A() as any;

func(a, 'a');
func(a, 'b');
// func(a, 'c'); // compile error

// func(ab, 'a'); // compile error
func(ab, 'b');

a.func('a');
a.func('b');
// a.func('c'); // compile error

// ab.func('a'); // compile error
ab.func('b'); // compile error NOT EXPECTED
```

### 🙁 Actual behavior

**Compiler Errors**

This expression is not callable.
  Each member of the union type '(<K extends keyof A>(k: K) => void) | (<K extends keyof B>(k: K) => void)' has signatures, but none of those signatures are compatible with each other.

### 🙂 Expected behavior

last line should not raise Compiler Errors, `(<K extends keyof A>(k: K) => void) | (<K extends keyof B>(k: K) => void)` should be union to (<K extends keyof A|B>(k: K) => void) ?

Also notice that global function "func" can work as expected.

**Comment by RyanCavanaugh** [comment]
Duplicate #33591, others.

> last line should not raise Compiler Errors, (<K extends keyof A>(k: K) => void) | (<K extends keyof B>(k: K) => void) should be union to (<K extends keyof A|B>(k: K) => void) ?

*In general* this is not a safe reduction step: contravariant uses of `K` would become unsound. It's very rarely possible to safely merge two arbitrary signatures.

**Comment by wenxiao** [comment]
@RyanCavanaugh I found a workaround here:

```ts
class BaseClass {
    func<K extends keyof this>(k: K) {}
}

class A extends BaseClass {
    a = 1;
    b = 2;
}

class B extends BaseClass {
    b = 3;
    c = 4;
}

function func<T, K extends keyof T>(v: T, k: K) {}

function workaroundFunc<T extends BaseClass, K extends keyof T>(v:T, k:K){
    v.func(k)
}

let a = new A();
let ab: A | B = new A() as any;

func(a, 'a');
func(a, 'b');
// func(a, 'c'); // compile error

// func(ab, 'a'); // compile error
func(ab, 'b');

a.func('a');
a.func('b');
// a.func('c'); // compile error

// ab.func('a'); // compile error
// ab.func('b'); // compile error NOT EXPECTED
// workaroundFunc(ab, 'a'); // compile error
workaroundFunc(ab, 'b');
```
[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBCkFMDC4rQN4FgBQ0-QDMBXAO2AB4BpaBADwBcESATGAawQE8B7A6egBYBLCAD4AFGwBc0SgEoMAXxzLsOUJBgBBGgyas4iFJow580MNAC80AIwBuM-gBG16ACZH2VetQxYuows-kZ+prgubgDMXubAbgAsXj7YxGT0QtwkhKQUACoANLKB+uxcvNB5EgBuMoXQ0rIK6ClpwBlZ0ADu3ABObGC93KTMAGK55HklwYYQyH5F1HRBBhw8fFXitfXS8lgReNUAdG2Scio4OCAI9BZuJAhd0Fricl7Xt2DOMjoAPnD3R7PV4WGBgEicLw4U5gIoAcjAcLe0Ny4lh0DhziRXgA9DicmQ0fDgNjoHjoMBuABbAAOQmuNF6Q16l2w5JhznhiLeZPxlNp9IQjOZKMJX3hWORamwYBOqIR2JwstOmMVbPxyvlJJ55P5dIZCCZfVZ5K+csJCp1fOp+qFhpF6osznNwHEqqtFJtguFfWgADkAPJTACiAA0AArBpB5YMAERw5J6-UGwxY4zFnIx3PsvM9AoNRpZ2CTAyGI3TrvFGMl9iAA)

by using `workaroundFunc`, I can do what I expected to do. I'm not sure what's the different between workaroundFunc and direct call.
